src/main/geometries/GeometryList.java

```diff
@@ -68,24 +78,47 @@ public void add(GeometryList geometries) {
 	 * all the geometries have been added, but before the ray tracing process begins.
 	 */
 	public void optimize() {
-		// TODO: implement
 		// move all finite geometries to a sub-geometries (use g.border().isFinite() to determine which ones to add)
+		GeometryList finiteGeometries = new GeometryList();
+		intersectibles.removeIf(i -> {
+			if(i.border().isFinite()) {
+				finiteGeometries.add(i);
+				return true;
+			}
+			return false;
+
+		});
 		finiteGeometries.constructHierarchy();
-                this.intersectibles.add(finiteGeometries);
+		this.intersectibles.add(finiteGeometries);
```

Comment:
This will just get flattened again, undoing all your hard work. You need to cast it to an Intersectible for it to go to the private add function which won't flatten it